modules: support tests/samples/boards in modules

nashif opened the pull request:
Generate options for sanitycheck to run tests and samples in modules.
Use the --sanitycheck-out <file> to generate a file that can be supplied
to sanitycheck on the commandline which will add additional testroots
and boards if the module does contain out of tree boards.

the module.yaml file now accepts the following:

samples:
  path: samples
tests:
  path: tests
boards:
  path: boards


- [x] Finalize `module.yml` structure
- [x] Documentation needed
- [ ] Test/sample module needed

nashif committed:
modules: support tests/samples/boards in modules

Generate options for sanitycheck to run tests and samples in modules.
Use the --sanitycheck-out <file> to generate a file that can be supplied
to sanitycheck on the commandline which will add additional testroots
and boards if the module does contain out of tree boards.

the module.yaml file now accepts the following:

samples:
  - path/to/samples
tests:
  - path/to/tests
boards:
  - path/to/boards

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

mbolivar commented on scripts/zephyr_module.py:
```diff
@@ -71,44 +89,79 @@ def process_module(module, cmake_out=None, kconfig_out=None):
                         help='File to write with resulting KConfig import'
                              'statements.')
+    parser.add_argument('--sanitycheck-out',
+                        help='File to write with resulting sanitycheck'
```
I think you're just copy pasting the style from other arguments, but there is a missing space.  It should be `sanitycheck '` instead.

mbolivar reviewed (commented):
Appreciate this! 

Can we get some corresponding documentation changes before merging this?

carlescufi reviewed (dismissed):
Thanks for the PR!
Can you add the relevant changes to:
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/zephyr_module.cmake#L30
as part of this PR?

carlescufi commented:
> Thanks for the PR!
> Can you add the relevant changes to:
> https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/zephyr_module.cmake#L30
> as part of this PR?

Actually, my mistake, `zephyr_module.cmake` is only for building samples, so I assume you won't do that here. But where are you going to invoke `zephyr_module.py --sanitycheck-out` ?

carlescufi commented on scripts/zephyr_module.py:
```diff
@@ -39,6 +39,27 @@
+      path:
+        type: str
+        required: false
+  boards:
```
Why did you add boards here, since that should really be part of the `sample.yaml` or `test.yaml`?

nashif commented:
> But where are you going to invoke `zephyr_module.py --sanitycheck-out` ?

in CI, generate the file and let sanitycheck loose on samples/tests in modules.

nashif committed:
doc: modules: document out-of-tree samples/tests

Document how to configure out-of-tree tests/samples and boards in the
module.yml file to be used by sanitycheck.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

tejlmand commented on doc/guides/modules.rst:
```diff
@@ -82,6 +82,26 @@ Each project in the ``west list`` output is tested like this:
+      tests:
+        path: tests
+      boards:
+        path: boards
```
Now, the existing configuration is actually implying a path:
`cmake: <path>` and `kconfig: <path/Kconfig-file>` under the build, so I kind of feel strange that `path` is a key by itself.

How about:
```suggestion
      samples: <path/to/samples>
      tests: <path/to/tests>
      boards: <path/to/boards>
```
where the path is relative from beginning of the module, which would most often be:
```
      samples: samples
      tests: tests
      boards: boards
```

Of course we might want an additional level besides `build`, as example `source` ? (feel free to provide better name)

carlescufi replied:
@tejlmand 
That's because you can have multiple `samples:` and `tests:` but you can't have multiple CMake or Kconfig files.

carlescufi replied:
@tejlmand @nashif @mbolivar my proposal would be to accept multiple entries for `path:`:

```
samples:
  path: path/to/sample/one
  paht: path/to/sample/two
```

carlescufi replied:
Or, as @nashif mentioned perhaps:

```
samples:
  - path/to/sample/one
  - path/to/sammple/two
```

mbolivar replied:
Doing it as a list is the right way YAML-wise.

carlescufi reviewed (changes requested):
You need this patch to make this work with MCUboot:

```
diff --git a/scripts/zephyr_module.py b/scripts/zephyr_module.py
index 1d15a87743..eeec936f23 100755
--- a/scripts/zephyr_module.py
+++ b/scripts/zephyr_module.py
@@ -35,7 +35,7 @@ METADATA_SCHEMA = '''
 type: map
 mapping:
   build:
-    required: true
+    required: false
     type: map
```

Other than that it seems to work fine, thank you!

carlescufi commented:
@mbolivar @tejlmand can you take another look and see if you are happy with the proposal?

carlescufi reviewed (approved):
Tested alongside:
https://github.com/zephyrproject-rtos/zephyr/pull/17868
and
https://github.com/JuulLabs-OSS/mcuboot/pull/619

mbolivar commented on doc/guides/modules.rst:
```diff
@@ -82,6 +82,26 @@ Each project in the ``west list`` output is tested like this:
        cmake: .
        kconfig: Kconfig
 
+- To execute both tests and samples available in modules the Zephyr test runner
```
s/samples in tests/samples and tests/

mbolivar commented on scripts/zephyr_module.py:
```diff
@@ -10,6 +10,11 @@
 Include file is generated for Kconfig using --kconfig-out.
 A <name>:<path> text file is generated for use with CMake using --cmake-out.
+
+Using --sanitycheck-out <filenam> an argument file for sanitycheck script will
```
filename

mbolivar commented on doc/guides/modules.rst:
```diff
@@ -82,6 +82,26 @@ Each project in the ``west list`` output is tested like this:
+        - boards
+
+  Additionally, if a module defines out of tree boards, the module file can
+  point sanitycheck to the path where those files are maintained in the module.
```
I think this ought to be before the code sample, no? It would be clearer.

mbolivar commented on scripts/zephyr_module.py:
```diff
@@ -10,6 +10,11 @@
+Using --sanitycheck-out <filenam> an argument file for sanitycheck script will
+be generated which would point to test and sample roots available in modules
+that can be included during a sanitycheck run. This allows testing code
+maintaiend in modules in addition to what is available in the main Zephyr tree.
```
Maintained

mbolivar reviewed (changes requested):
Minor tweaks to docs suggested.

I wish this could be made simpler, to be honest, but I'm in favor of getting the feature in and iterating as needed.

nashif committed:
ci: run sanitycheck on tests and samples

If modules announce they have tests, run sanitycheck on those modules.
This will run on first matrix node and as the last step after running
regular testsuite.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
